include/faabric/scheduler/MpiWorld.h

```diff
@@ -3,19 +3,41 @@
 #include <faabric/mpi/mpi.h>
 
 #include <faabric/proto/faabric.pb.h>
-#include <faabric/scheduler/FunctionCallClient.h>
 #include <faabric/scheduler/InMemoryMessageQueue.h>
-#include <faabric/scheduler/MpiThreadPool.h>
-#include <faabric/state/StateKeyValue.h>
+#include <faabric/transport/MpiMessageEndpoint.h>
+#include <faabric/util/logging.h>
+#include <faabric/util/timing.h>
 
 #include <atomic>
-#include <thread>
+#include <list>
 
 namespace faabric::scheduler {
 typedef faabric::util::Queue<std::shared_ptr<faabric::MPIMessage>>
   InMemoryMpiQueue;
 
-std::string getWorldStateKey(int worldId);
+/* The untracked message buffer (UMB) keeps track of the asyncrhonous
+ * message that we must have received (i.e. through an irecv call) but we
+ * still have not waited on (acknowledged). Messages are acknowledged either
+ * through a call to recv or a call to await. A call to recv will
+ * acknowledge (i.e. synchronously read from transport buffers) as many
+ * unacknowleged messages there are, plus one.
+ */
+struct UnackedMessageBuffer
+{
+    struct Arguments
+    {
+        int sendRank;
+        int recvRank;
+        uint8_t* buffer;
+        faabric_datatype_t* dataType;
+        int count;
+        faabric::MPIMessage::MPIMessageType messageType;
+    };
```

Comment:
If we put the shared pointer to the message into this struct, could we get rid of one of the `std::lists` in this class?

Reply:
Yes, in fact after some deep refactoring I realized we could just as well use one single list (which makes iterator management much easier).